### PR TITLE
Update developer documentation to provide extra ANTLR instructions

### PR DESCRIPTION
**Pull request by jimidle**
Adds additional instructions for making ANTLR grammar changes, including explicit rules and checks to make.

### CODE REVIEW

File: core/README.md [hunk 1]
```diff
@@ -19,14 +19,38 @@ Changing the ANTLR grammars is a specialized task, and only a few team members h
...
+out of almost any input specification. However, the resulting parser may then be very slow, 
 or may not be able to handle certain edge cases.
 
-After a grammar change is made, the .g4 files must be reformatted to stay in line with the guidelines. We use (for now
-at least) the [antlr-format](https://github.com/mike-lischke/antlr-format) tool to do this. The tool was written
-in typescript, so it does not easily run as a maven task. For the moment, please run it manually on the .g4 files
-in each of the dialects we support under `src/main/antlr4/../parsers`
+After a grammar change is made, the .g4 files must be reformatted to stay in line with the guidelines. 
+We use (for now at least) the [antlr-format](https://github.com/mike-lischke/antlr-format) tool to do this. The tool is run as part of the 
+maven build process, if you are using the 'format' maven profile.
```
Comment (nfx) [inline core/README.md:29]:
```suggestion
maven build process, if you are using the 'format' maven profile or through `make fmt`  or `make fmt-scala`.
```

File: core/README.md [hunk 2]
```diff
@@ -19,14 +19,38 @@ Changing the ANTLR grammars is a specialized task, and only a few team members h
...
+
+## Checking ANTLR changes
+
+If you make changes to the ANTLR grammar, you should check the following things 
+(this is a good order to check them in):
+
+ - Have you defined any new TOKENS used in the PARSER?
+   - If so, did you define them in the LEXER?
```
Comment (nfx) [inline core/README.md:37]: add before/after examples for this to be more friendly to junior coders.

File: core/README.md [hunk 3]
```diff
@@ -19,14 +19,38 @@ Changing the ANTLR grammars is a specialized task, and only a few team members h
...
+ - Have you eliminated the use of any tokens (this is generally a good thing - such as replacing 
+   a long list of option keywords with an id rule, which is then checked later in the process)?
+   - If so, have you removed them from the lexer?
+ - Have you added any new rules? If so:
+   - have you checked that you have not duplicated some syntactic structure that is already defined
+     and could be reused? 
+   - have you checked that they are used in the parser (the IntelliJ plugin will highlight unused rules)?
```
Comment (nfx) [inline core/README.md:44]: add screenshot :)

File: core/README.md [hunk 4]
```diff
@@ -19,14 +19,38 @@ Changing the ANTLR grammars is a specialized task, and only a few team members h
...
+If you make changes to the ANTLR grammar, you should check the following things 
+(this is a good order to check them in):
+
+ - Have you defined any new TOKENS used in the PARSER?
+   - If so, did you define them in the LEXER?
+ - Have you eliminated the use of any tokens (this is generally a good thing - such as replacing 
+   a long list of option keywords with an id rule, which is then checked later in the process)?
+   - If so, have you removed them from the lexer?
+ - Have you added any new rules? If so:
+   - have you checked that you have not duplicated some syntactic structure that is already defined
```
Comment (nfx) [inline core/README.md:42]: add good/bad example

**Review by nfx — CHANGES_REQUESTED** [review]

**Comment by jimidle** [comment]
Docs greatly expanded with examples and screenshots, as well as explanations of various parts of grammar files.

**Review by nfx — APPROVED** [review]
thank you!

**Pull request merged**